src/garage/torch/policies/context_conditioned_policy.py

```diff
@@ -0,0 +1,283 @@
+        no = torch.as_tensor(timestep.next_observation[None, None, ...],
+                             device=tu.global_device()).float()
+
+        if self._use_next_obs:
+            data = torch.cat([o, a, r, no], dim=2)
+        else:
+            data = torch.cat([o, a, r], dim=2)
+
+        if self._context is None:
+            self._context = data
+        else:
+            self._context = torch.cat([self._context, data], dim=1)
+
+    # pylint: disable=attribute-defined-outside-init
+    def infer_posterior(self, context):
+        r"""Compute :math:`q(z \| c)` as a function of input context and sample new z.
+
+        Args:
+            context (torch.Tensor): Context values, with shape
+                :math:`(X, N, C)`. X is the number of tasks. C is the
+                combined size of observation, action, reward, and next
+                observation if next observation is used in context. Otherwise,
+                C is the combined size of observation, action, and reward.
+
+        """
+        params = self._context_encoder(context)
+        # pylint: disable=protected-access
+        params = params.view(context.size(0), -1,
+                             self._context_encoder._output_dim)
+        # with probabilistic z, predict mean and variance of q(z | c)
+        if self._use_information_bottleneck:
+            mu = params[..., :self._latent_dim]
+            sigma_squared = F.softplus(params[..., self._latent_dim:])
+            z_params = [
+                tu.product_of_gaussians(m, s)
+                for m, s in zip(torch.unbind(mu), torch.unbind(sigma_squared))
+            ]
+            self.z_means = torch.stack([p[0] for p in z_params])
+            self.z_vars = torch.stack([p[1] for p in z_params])
+        else:
+            self.z_means = torch.mean(params, dim=1)
+        self.sample_from_belief()
+
+    # pylint: disable=arguments-differ
+    def forward(self, obs, context):
+        """Given observations and context, get actions and probs from policy.
+
+        Args:
+            obs (torch.Tensor): Observation values, with shape
+                :math:`(X, N, O)`. X is the number of tasks.
+            context (torch.Tensor): Context values, with shape
+                :math:`(X, N, C)`. C is the combined size of observation,
+                action, reward, and next observation if next observation is
+                used in context. Otherwise, C is the combined size of
+                observation, action, and reward.
+
+        Returns:
+            tuple:
+                * torch.Tensor: Predicted action values.
+                * np.ndarray: Mean of distribution.
+                * np.ndarray: Log std of distribution.
+                * torch.Tensor: Log likelihood of distribution.
+                * torch.Tensor: Sampled values from distribution before
+                    applying tanh transformation.
+            torch.Tensor: z values, with shape :math:`(N, latent_dim)`.
```

Comment:
please use only symbols, not text, in math environments.

are the z values the most frequently used output here? if so, it would be most pythonic for z to be the first output, and the others to be packed into the second tuple, e.g.

```python
z, _ = foo.forward()
```

In fact, how many of these outputs are actually used? If one of them isn't just remove it from the returns.